storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java

```diff
@@ -45,54 +47,68 @@
-  private volatile Bytes32 bestBlockRoot =
-      Bytes32.ZERO; // block chosen by lmd ghost to build and attest on
+  private volatile Optional<Bytes32> bestBlockRoot =
+      Optional.empty(); // block chosen by lmd ghost to build and attest on
   private volatile UnsignedLong bestSlot =
       UnsignedLong.ZERO; // slot of the block chosen by lmd ghost to build and attest on
   // Time
   private volatile UnsignedLong genesisTime;
 
   public static ChainStorageClient memoryOnlyClient(final EventBus eventBus) {
-    return new ChainStorageClient(eventBus, TransactionPrecommit.memoryOnly());
+    final ChainStorageClient client =
+        new ChainStorageClient(eventBus, TransactionPrecommit.memoryOnly());
+    eventBus.register(client);
+    return client;
+  }
+
+  public static SafeFuture<ChainStorageClient> storageBackedClient(final EventBus eventBus) {
+    final StorageBackedChainStorageClientFactory factory =
+        new StorageBackedChainStorageClientFactory(eventBus);
+    eventBus.register(factory);
+    return factory.get();
   }
 
-  public static ChainStorageClient storageBackedClient(final EventBus eventBus) {
-    return new ChainStorageClient(eventBus, TransactionPrecommit.storageEnabled(eventBus));
+  @VisibleForTesting
+  static ChainStorageClient memoryOnlyClientWithStore(final EventBus eventBus, final Store store) {
+    final ChainStorageClient client =
+        new ChainStorageClient(eventBus, TransactionPrecommit.memoryOnly());
+    eventBus.register(client);
+    client.setStore(store);
+    return client;
   }
 
-  private ChainStorageClient(EventBus eventBus, final TransactionPrecommit transactionPrecommit) {
+  ChainStorageClient(EventBus eventBus, final TransactionPrecommit transactionPrecommit) {
     this.eventBus = eventBus;
     this.transactionPrecommit = transactionPrecommit;
-    this.eventBus.register(this);
   }
 
-  public void initializeFromGenesis(final BeaconState initialState) {
-    setGenesisTime(initialState.getGenesis_time());
-    final Store store = Store.get_genesis_store(initialState);
-    setStore(store);
-    eventBus.post(new StoreGenesisDiskUpdateEvent(store));
+  public void subscribeStoreInitialized(Runnable runnable) {
+    storeInitializedFuture.always(runnable);
+  }
 
-    // The genesis state is by definition finalised so just get the root from there.
-    Bytes32 headBlockRoot = store.getFinalizedCheckpoint().getRoot();
-    BeaconBlock headBlock = store.getBlock(headBlockRoot);
-    updateBestBlock(headBlockRoot, headBlock.getSlot());
-    eventBus.post(new StoreInitializedEvent());
+  public void subscribeBestBlockInitialized(Runnable runnable) {
+    bestBlockInitialized.always(runnable);
   }
 
-  public void initializeFromStore(final Store store, final Bytes32 headBlockRoot) {
-    BeaconState state = store.getBlockState(headBlockRoot);
-    setGenesisTime(state.getGenesis_time());
-    setStore(store);
+  public void setGenesisState(final BeaconState genesisState) {
```

Comment:
Not a big deal but I think I preferred the `initializeFromGenesis` naming because it doesn't feel like it quite matches the pattern of a setter (particularly that it can't be called multiple times).

Reply:
Fair enough - will switch it back